### DISCUSSION TITLE
Manual triggered recipe web hook

### DISCUSSION THREAD

**Discussion by m42e** [mealie-recipes/mealie#1036] [Feature Request]
**Before submitting this feature request I have**
- [x] Reviewed the documentation and verified that my feature request isn't already a feature.
- [x] Reviewed the existing feature requests and my feature has not been requested.
- [x] Checked the [tasks taged](https://github.com/hay-kot/mealie/issues?q=is%3Aissue+is%3Aopen+label%3Atask+) issues and verified my feature is not covered

**Please Describe The Problem To Be Solved**
I have already some workflow in place, to add stuff to my (old school, printed) shopping list. Additionally I would like to utilise the recipe database a bit more and try out other automation stuff. E.g. Printing the recipe on a thermal printer, ect.

**Suggest A Solution**
Allowing to trigger a (specific) web hook for a recipe manually from the recipe page would be a good way of triggering a workflow that is outside the boundaries of mealie and not to have to copy around urls or other stuff.
- Button on the recipe page to trigger the hook
- configuration for the hook
- triggering of the hook within the backend

**Additional Information**
- [x] If the feature is accepted I'm willing to submit a PR to provide this feature

**Comment by hay-kot** [comment]
I'm open to this, sounds like it's something you want to help out with, Would you mind fleshing out the proposal a bit and outlining how it would be setup and used by a mealie user?

**Reply by m42e** [reply]
Sure I will have a look into it.

**Comment by m42e** [comment]
# User Interface

## Configuration

- The user can configure a custom url as webhook. This setting is user specific. 
- Additionally an option to include the recipe information or only the url. 
- Probably we want an additional option to disable SSL verfication.
- Configureable POST/GET
- Show detailed result of last call. (error and success)

## Triggering

- Besides the UI elements to share or print a recipe an additional button will be added to trigger the web hook for the current shown recipe

# Backend

- The backend performs the trigger of the web hook asynchronously when triggered from the user
- The backend logs possible error and success and provides the last result, including returned content.
- There are no retries.

**Reply by hay-kot** [reply]
I think this is a pretty good implementation, I think it's enough to get started on a PR - but I think this one is going to take some iterations to get right.

### Questions / Comments

> The user can configure a custom url as webhook. This setting is user specific.

Can only 1 web-hook be created per user? I think multiple would be preferred. 

> Show detailed result of last call. (error and success)

Where would we log this, just have a field in the database? Maybe we should provide a way to test the webhook in the UI? 

> Besides the UI elements to share or print a recipe an additional button will be added to trigger the web hook for the current shown recipe

Not super important at this stage, but I think I'd stick this under a "More" button so we don't keep cluttering up that Menu, I've got a few other things I want to add to it later down the line.

> The backend logs possible error and success and provides the last result, including returned content.

Log to console or log somewhere else?

**Comment by michael-genson** [comment]
This is now supported with [Recipe Actions](https://github.com/mealie-recipes/mealie/pull/3448)
